lib/main.dart

```diff
@@ -0,0 +1,189 @@
+    String version;
+    String channel;
+
+    if (sdk == 'stable' || sdk == 'beta' || sdk == 'dev') {
+      channel = sdk;
+      version =
+          raw ? 'latest' : (await latestPublishedVersion(channel, flavor))!;
+    } else if (sdk == 'main') {
+      channel = 'be';
+      version =
+          raw ? 'latest' : (await latestPublishedVersion(channel, flavor))!;
+    } else {
+      version = sdk;
+
+      // Derive the channel from the version string.
+      if (sdk.contains('dev')) {
+        channel = 'dev';
+      } else if (sdk.contains('beta')) {
+        channel = 'beta';
+      } else if (sdk.contains('main')) {
+        core.setFailed('Versions cannot be specified for main channel builds.');
+        return;
+      } else {
+        channel = 'stable';
+      }
+    }
+
+    core.info('Installing the $os-$architecture Dart SDK version $version from '
+        'the $channel ($flavor) channel.');
+
+    // Calculate url based on https://dart.dev/tools/sdk/archive#download-urls.
+    final url = 'https://storage.googleapis.com/dart-archive/'
+        'channels/$channel/$flavor/$version/sdk/'
+        'dartsdk-$os-$architecture-release.zip';
+
+    // Use a cached sdk or download and cache the sdk; using a 'raw' sdk flavor
+    // disables caching.
+    final toolName = flavor == 'raw' ? 'dart_raw' : 'dart';
+    var sdkPath = !raw ? toolCache.find(toolName, version, architecture) : '';
+    if (sdkPath.isNotEmpty) {
+      core.info('Using cached sdk from $sdkPath.');
+    } else {
+      core.info('$url ...');
+
+      final archivePath =
+          await promiseToFuture<String>(toolCache.downloadTool(url));
+      var extractedFolder =
+          await promiseToFuture<String>(toolCache.extractZip(archivePath));
+      extractedFolder = path.join(extractedFolder, 'dart-sdk');
+
+      sdkPath = await promiseToFuture<String>(
+          toolCache.cacheDir(extractedFolder, toolName, version, architecture));
+    }
+
+    final pubCache = path.join(
+        getProcessEnv(os == 'windows' ? 'USERPROFILE' : 'HOME')!, '.pub-cache');
+
+    core.exportVariable('DART_HOME', sdkPath);
+    core.addPath(path.join(sdkPath, 'bin'));
+    core.exportVariable('PUB_CACHE', pubCache);
+    core.addPath(path.join(pubCache, 'bin'));
+
+    // Create the OIDC token used for pub.dev publishing.
+    await createPubOIDCToken();
+
+    // Configure the outputs.
+    if (raw) {
+      core.setOutput('dart-version', getVersionFromSdk(sdkPath)!);
+    } else {
+      core.setOutput('dart-version', version);
+    }
+
+    // Report success; print version.
+    await promiseToFuture(exec.exec('dart', ['--version']));
+  } catch (error) {
+    core.setFailed('$error');
+  }
+}
+
+String? getVersionFromSdk(String sdkPath) {
+  final versionFilePath = path.join(sdkPath, 'version');
+  print(versionFilePath);
+
+  if (fs.existsSync(versionFilePath)) {
+    return fs.readFileSync(versionFilePath, 'utf8').trim();
+  } else {
+    return null;
```

Comment:
Given that the one place that calls this asserts not-null, maybe this should be non-nullable instead and unconditionally read the file? We still might get random IO related errors from fs.readFileSync and the file doesn't exist we get a runtime error anyway.

Reply:
sounds good; makes sense to not expect a missing file